File: app/src/main/java/com/remotearthsolutions/expensetracker/fragments/AddCategoryDialogFragment.java
package com.remotearthsolutions.expensetracker.fragments;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.*;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.remotearthsolutions.expensetracker.R;
import com.remotearthsolutions.expensetracker.adapters.IconListAdapter;
import com.remotearthsolutions.expensetracker.databaseutils.DatabaseClient;
import com.remotearthsolutions.expensetracker.databaseutils.daos.CategoryDao;
import com.remotearthsolutions.expensetracker.databaseutils.models.CategoryModel;
import com.remotearthsolutions.expensetracker.utils.CategoryIcons;
import com.remotearthsolutions.expensetracker.utils.Constants;
import com.remotearthsolutions.expensetracker.utils.Utils;
import io.reactivex.Completable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

import java.util.List;

public class AddCategoryDialogFragment extends DialogFragment {


    private IconListAdapter iconListAdapter;
    private RecyclerView recyclerView;
    private AddCategoryDialogFragment.Callback callback;
    private EditText categoryNameEdtxt;
    private TextView categorydialogstatus;
    private CategoryModel categoryModel;
    private String selectedIcon;
    private Context context;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.context = context;
    }

    public AddCategoryDialogFragment() {
    }


    public static AddCategoryDialogFragment newInstance(String title) {
        AddCategoryDialogFragment frag = new AddCategoryDialogFragment();
        Bundle args = new Bundle();
        args.putString(Constants.KEY_TITLE, title);
        frag.setArguments(args);
        return frag;
    }

    public void setCallback(AddCategoryDialogFragment.Callback callback) {
        this.callback = callback;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_add_update_category_account, container);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        categoryNameEdtxt = view.findViewById(R.id.nameEdtxt);
        categorydialogstatus = view.findViewById(R.id.header);
        Button okBtn = view.findViewById(R.id.okBtn);

        if (categoryModel != null) {
            categoryNameEdtxt.setText(categoryModel.getName());
            categoryNameEdtxt.setSelection(categoryNameEdtxt.getText().length());
            categorydialogstatus.setText(getString(R.string.update_category));
            okBtn.setText(getString(R.string.update));
        }

        okBtn.setOnClickListener(v -> saveCategory());

        recyclerView = view.findViewById(R.id.accountrecyclearView);
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                (Utils.getDeviceScreenSize(context).height / 2));
        recyclerView.setLayoutParams(params);
        recyclerView.setHasFixedSize(true);
        GridLayoutManager gridLayoutManager = new GridLayoutManager(context, 4);
        recyclerView.setLayoutManager(gridLayoutManager);

        List<String> alliconList = CategoryIcons.getAllIcons();
        iconListAdapter = new IconListAdapter(alliconList, gridLayoutManager);
        iconListAdapter.setSelectedIcon(selectedIcon != null ? selectedIcon : "");
        iconListAdapter.setOnItemClickListener(icon -> {
            selectedIcon = icon;
            iconListAdapter.setSelectedIcon(selectedIcon != null ? selectedIcon : "");
            iconListAdapter.notifyDataSetChanged();
        });
        recyclerView.setAdapter(iconListAdapter);

    }

    private void saveCategory() {

        final String categoryName = categoryNameEdtxt.getText().toString().trim();

        if (categoryName.isEmpty()) {
            categoryNameEdtxt.setError(getString(R.string.enter_a_name));
            categoryNameEdtxt.requestFocus();
            return;
        }

        if (selectedIcon == null || selectedIcon.isEmpty()) {
            Toast.makeText(getActivity(), getString(R.string.select_an_icon), Toast.LENGTH_SHORT).show();
            return;
        }

        CategoryDao categoryDao = DatabaseClient.getInstance(getContext()).getAppDatabase().categoryDao();

        if (categoryModel == null) {
            categoryModel = new CategoryModel();
        }
        categoryModel.setName(categoryName);
        categoryModel.setIcon(selectedIcon);

        CompositeDisposable compositeDisposable = new CompositeDisposable();
        compositeDisposable.add(Completable.fromAction(() -> {
            if (categoryModel.getId() > 0) {
                categoryDao.updateCategory(categoryModel);
            } else {
                categoryDao.addCategory(categoryModel);
            }
        }).subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(() -> callback.onCategoryAdded(categoryModel)));

    }

    public void setCategory(CategoryModel categoryModel) {
        this.categoryModel = categoryModel;
        if (categoryModel != null) {
            selectedIcon = categoryModel.getIcon();
        }

    }

    public interface Callback {
        void onCategoryAdded(CategoryModel categoryModel);
    }


}

File: app/src/main/java/com/remotearthsolutions/expensetracker/fragments/AddCategoryDialogFragment.kt
package com.remotearthsolutions.expensetracker.fragments

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.Toast
import androidx.fragment.app.DialogFragment
import androidx.recyclerview.widget.GridLayoutManager
import com.remotearthsolutions.expensetracker.R
import com.remotearthsolutions.expensetracker.adapters.IconListAdapter
import com.remotearthsolutions.expensetracker.databaseutils.DatabaseClient
import com.remotearthsolutions.expensetracker.databaseutils.models.CategoryModel
import com.remotearthsolutions.expensetracker.databinding.FragmentAddUpdateCategoryAccountBinding
import com.remotearthsolutions.expensetracker.utils.AnalyticsManager
import com.remotearthsolutions.expensetracker.utils.CategoryIcons.allIcons
import com.remotearthsolutions.expensetracker.utils.Constants
import com.remotearthsolutions.expensetracker.utils.Utils.getDeviceScreenSize
import io.reactivex.Completable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

class AddCategoryDialogFragment : DialogFragment() {
    private lateinit var binding: FragmentAddUpdateCategoryAccountBinding
    private lateinit var iconListAdapter: IconListAdapter
    private var callback: Callback? = null
    private var categoryModel: CategoryModel? = null
    private var selectedIcon: String? = null
    private lateinit var mContext: Context
    override fun onAttach(context: Context) {
        super.onAttach(context)
        mContext = context
    }

    fun setCallback(callback: Callback?) {
        this.callback = callback
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentAddUpdateCategoryAccountBinding.inflate(layoutInflater, container, false)
        return binding.root
    }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?
    ) {
        super.onViewCreated(view, savedInstanceState)
        if (categoryModel != null) {
            binding.nameEdtxt.setText(categoryModel!!.name)
            binding.nameEdtxt.setSelection(binding.nameEdtxt.text.length)
            binding.header.text = requireContext().getString(R.string.update_category)
            binding.okBtn.text = requireContext().getString(R.string.update)
        }
        binding.okBtn.setOnClickListener { saveCategory() }
        val params = LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            getDeviceScreenSize(mContext)!!.height / 2
        )
        binding.accountrecyclearView.layoutParams = params
        binding.accountrecyclearView.setHasFixedSize(true)
        val gridLayoutManager = GridLayoutManager(mContext, 4)
        binding.accountrecyclearView.layoutManager = gridLayoutManager
        val allIconList = allIcons
        iconListAdapter = IconListAdapter(allIconList, gridLayoutManager)
        iconListAdapter.setSelectedIcon(if (selectedIcon != null) selectedIcon else "")
        iconListAdapter.setOnItemClickListener(object : IconListAdapter.OnItemClickListener {
            override fun onItemClick(icon: String?) {
                selectedIcon = icon
                iconListAdapter.setSelectedIcon(if (selectedIcon != null) selectedIcon else "")
                iconListAdapter.notifyDataSetChanged()
            }

        })
        binding.accountrecyclearView.adapter = iconListAdapter
    }

    private fun saveCategory() {
        val categoryName = binding.nameEdtxt.text.toString().trim { it <= ' ' }
        if (categoryName.isEmpty()) {
            binding.nameEdtxt.error = requireContext().getString(R.string.enter_a_name)
            binding.nameEdtxt.requestFocus()
            return
        }
        if (categoryName.length > 20) {
            binding.nameEdtxt.error =
                requireContext().getString(R.string.name_should_be_less_than_20_char)
            binding.nameEdtxt.requestFocus()
            return
        }
        if (selectedIcon == null || selectedIcon!!.isEmpty()) {
            Toast.makeText(
                activity,
                requireContext().getString(R.string.select_an_icon),
                Toast.LENGTH_SHORT
            )
                .show()
            return
        }
        val categoryDao =
            DatabaseClient.getInstance(mContext).appDatabase.categoryDao()
        if (categoryModel == null) {
            categoryModel = CategoryModel()
        }
        categoryModel!!.name = categoryName
        categoryModel!!.icon = selectedIcon
        val compositeDisposable = CompositeDisposable()
        compositeDisposable.add(Completable.fromAction {
            if (categoryModel!!.id > 0) {
                categoryDao.updateCategory(categoryModel)
                with(AnalyticsManager) {
                    logEvent(EXPENSE_CATEGORY_UPDATED)
                }
            } else {
                categoryDao.addCategory(categoryModel!!)
                with(AnalyticsManager) {
                    logEvent(EXPENSE_CATEGORY_CREATED)
                }
            }
        }.subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { callback!!.onCategoryAdded(categoryModel) }
        )
    }

    fun setCategory(categoryModel: CategoryModel?) {
        this.categoryModel = categoryModel
        if (categoryModel != null) {
            selectedIcon = categoryModel.icon
        }
    }

    interface Callback {
        fun onCategoryAdded(categoryModel: CategoryModel?)
    }

    companion object {
        fun newInstance(title: String?): AddCategoryDialogFragment {
            val frag = AddCategoryDialogFragment()
            val args = Bundle()
            args.putString(Constants.KEY_TITLE, title)
            frag.arguments = args
            return frag
        }
    }
}